### ISSUE TITLE
ExceptionNameCheck

### ISSUE THREAD

**Issue by george-007** [checkstyle/checkstyle#3710]
It would be nice  to have a check for exception naming, i.e. classes that extends ``^.*Exception$``  has suffix ``Exception``.

**Comment by romani** [comment]
Check need to be generalized to work for other types. 
if extended class with some pattern , keep pattern (extends "*Error" should keep name as "*Error")

**Comment by george-007** [comment]
What about two patterns then:
if extending class with name that matches ``pattern1``, class name should match ``pattern2``.

**Comment by romani** [comment]
Yes, we need to figureout good name for that that could be clear for user witjout reading documentation

**Comment by jochenvdv** [comment]
I propose to name the check `InheritedTypeName`.
For the properties I suggest `supertypeNamePattern `and `subtypeNamePattern`.

I prefer `Type ` over `Class ` because this check should also be usable on interfaces.

**Comment by romani** [comment]
Should it treat "extents" and "implements" as the same trigger for activation ?

**Comment by jochenvdv** [comment]
At least `extends`.

Example which allows `SortedSet `extending `Set`, but not `SetThatIsSorted`:

```xml
<module name="InheritedTypeName">
    <property name="supertypeNamePattern" value="^Set$"/>
    <property name="subtypeNamePattern" value="^.*Set$"/>
</module>
```

Working with `implements` will require more thought since multiple interfaces can be implemented and configurations could conflict.

Example that cannot be satisfied by subtype (name cannot end in both `InterfaceA `and `InterfaceB`):

```xml
<module name="InheritedTypeName">
    <property name="supertypeNamePattern" value="^InterfaceA$"/>
    <property name="subtypeNamePattern" value="^.*InterfaceA$"/>
</module>
<module name="InheritedTypeName">
    <property name="supertypeNamePattern" value="^InterfaceB$"/>
    <property name="subtypeNamePattern" value="^.*InterfaceB$"/>
</module>
```

Example which allows a class implementing `InterfaceA `and `InterfaceB `to be named `ClassImplementingInterfaceAInterfaceB`:

```xml
<module name="InheritedTypeName">
    <property name="supertypeNamePattern" value="^InterfaceA$"/>
    <property name="subtypeNamePattern" value="^.*InterfaceA.*$"/>
</module>
<module name="InheritedTypeName">
    <property name="supertypeNamePattern" value="^InterfaceB$"/>
    <property name="subtypeNamePattern" value="^.*InterfaceB.*$"/>
</module>
```

Let me know what you think. We could create an extra property which defines if the check is to be triggered by `extends` and/or `implements`.

**Comment by rnveach** [comment]
> Should it treat "extents" and "implements" as the same trigger for activation ?

And how are you going to handle a class that implements multiple interfaces, each one with their own naming convention? Even a single `extend` and `implement` could have conflicting naming conventions.

As just an example, I ran into something sorta similar with a custom check I just made, but it uses `AbstractTypeAwareCheck` to look at the entire class hierarchy. In my case, I had classes extending `ApplicationUserData` which extended `UserData`, both of which had their own naming convention, `User$` and `UserData$` respectively. In my case, I made the top hierarchy of the class have precedence.

Top hierarchy won't work in this case because we can only look at the one class by itself, and I assume each interface will get its own isolated check and not be one big list like `CustomDeclarationOrder`.
